operator/pkg/clustermesh/client.go

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package clustermesh
+
+import (
+	"context"
+
+	discovery "k8s.io/api/discovery/v1"
+	v1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	discoveryv1 "k8s.io/client-go/kubernetes/typed/discovery/v1"
+	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+const (
+	meshRealServiceNameLabel = "mesh.cilium.io/real-service-name"
+)
+
+// meshClient override a few EndpointSlice methods to add/remove "hacks" on EndpointSlice.
+// This is necessary because the service informer passed to the controller will create
+// fake servies with this form `$svcName-$clusterName`. So this client will fix back
+// the service label name to the correct service.
+type meshClient struct {
+	k8sClient.Clientset
+}
+
+func (c meshClient) DiscoveryV1() discoveryv1.DiscoveryV1Interface {
+	return meshClientDiscoveryV1{c.Clientset.DiscoveryV1()}
+}
+
+type meshClientDiscoveryV1 struct {
+	discoveryv1.DiscoveryV1Interface
+}
+
+func (c meshClientDiscoveryV1) EndpointSlices(namespace string) discoveryv1.EndpointSliceInterface {
+	return meshClientEndpointSlice{c.DiscoveryV1Interface.EndpointSlices(namespace)}
+}
+
+type meshClientEndpointSlice struct {
+	discoveryv1.EndpointSliceInterface
+}
+
+func isServiceOwnerReferrence(ownerReference metav1.OwnerReference) bool {
+	return ownerReference.APIVersion == "v1" && ownerReference.Kind == "Service"
+}
+
+// addEndpointSliceMeshHacks fakes that the parent service of the EndpointSlice
+// has the cluser name in its name
+func addEndpointSliceMeshHacks(endpointSlice *v1.EndpointSlice) {
+	if endpointSlice == nil || endpointSlice.Labels[discovery.LabelManagedBy] != controllerName {
+		return
+	}
+
+	endpointSlice.Labels[meshRealServiceNameLabel] = endpointSlice.Labels[discovery.LabelServiceName]
+	endpointSlice.Labels[discovery.LabelServiceName] = endpointSlice.Labels[meshRealServiceNameLabel] + "-" + endpointSlice.Labels[mcsapiv1alpha1.LabelSourceCluster]
```

Comment:
I expanded a bit on those things in the existing comment in this file https://github.com/cilium/cilium/compare/f3d734ea321f25c46a5405437937d83e6ce1a0a3..74c9bb927c7657a51bb4ca85a1374507e7e85e72 
and there is also some bits here https://github.com/MrFreezeex/cilium/blob/74c9bb927c7657a51bb4ca85a1374507e7e85e72/operator/pkg/clustermesh/service_informer.go#L32

So the service are sent with the cluster name as suffix and the point of this client fix it back when writing (and put back this kind of hack on reading to prevent issues when reconciling back).

Reply:
I meant a comment in the code, not in the GH UI. Sorry if I wasn't clear with that :sweat_smile:

Reply:
The second paragraph was more a TLDR but as I was pointing out there is some comments in the code already about that line 25 and line 56 in this file sorry if that wasn't clear